rtt_ros2_params/src/rtt_ros2_params.cpp

```diff
@@ -0,0 +1,296 @@
+{
+
+Params::Params(RTT::TaskContext * owner)
+: RTT::Service("Params", owner)
+{
+  RTT::Logger::In in(getName());
+
+  RTT::log(RTT::Info) <<
+    "Parameter service instantiated! in " << getName() <<
+    RTT::endlog();
+
+  RTT::log(RTT::Debug) <<
+    "Name of the owner: " << owner->getName() <<
+    getName() << RTT::endlog();
+  const auto check = owner->provides()->hasService("ros");
+  RTT::log(RTT::Debug) <<
+    "ROS2 exists: " << check << RTT::endlog();
+
+  this->doc(
+    "RTT Service for synchronizing ROS2 parameters with the properties of a corresponding "
+    "RTT component");
+
+  addOperation("getParameter", &Params::getParameter, this, RTT::ClientThread)
+  .doc("Gets a parameter from the node parameter server")
+  .arg("name", "Name of the parameter to retrieve");
+
+  addOperation("setParameter", &Params::setParameter, this, RTT::ClientThread)
+  .doc("Sets a parameter to the node parameter server")
+  .arg("name", "Name of the parameter to retrieve")
+  .arg("value", "The value of the parameter to set");
+
+  addOperation("loadProperty", &Params::loadProperty, this, RTT::ClientThread)
+  .doc("Loads a parameter from the node parameter server into a Property")
+  .arg("parameter_name", "Name of the parameter to retrieve")
+  .arg("property_name", "Name of the property to load into");
+
+  addOperation("storeProperty", &Params::storeProperty, this, RTT::ClientThread)
+  .doc("Stores a property into a node parameter server")
+  .arg("property_name", "Name of the property to store")
+  .arg("parameter_name", "Name of the parameter to store into");
+}
+
+Params::~Params() = default;
+
+bool Params::check_ros2_node_in_component()
+{
+  return getOwner()->provides()->hasService("Node");
+}
+
+bool Params::check_ros2_node_in_global()
+{
+  return RTT::internal::GlobalService::Instance()->hasService("ros");
+}
+
+rclcpp::ParameterValue Params::getParameter(const std::string param_name)
+{
+  RTT::Logger::In in(getName());
+
+  RTT::log(RTT::Debug) <<
+    "Retrieving the parameter \"" << param_name << "\"" <<
+    RTT::endlog();
+
+  rclcpp::ParameterValue paramvalue;
+
+  rclcpp::Node::SharedPtr rosnode;
+  if (get_ros2_node(rosnode)) {
+    rosnode->get_parameter(param_name, paramvalue);
+  } else {
+    RTT::log(RTT::Error) <<
+      "The ROS2 node doesn't exist, no parameter can be retrieved. Import rtt_ros2_node first." <<
+      RTT::endlog();
+  }
+
+  return paramvalue;
+}
+
+bool Params::setParameter(const std::string param_name, const rclcpp::ParameterValue paramvalue)
+{
+  RTT::Logger::In in(getName());
+
+  RTT::log(RTT::Debug) <<
+    "Setting the parameter \"" << param_name << "\"" <<
+    RTT::endlog();
+
+  rclcpp::Node::SharedPtr rosnode;
+  if (get_ros2_node(rosnode)) {
+    if (!rosnode->has_parameter(param_name)) {
+      // Create new parameter
+      RTT::log(RTT::Warning) <<
+        "The parameter did not exist, creating one" <<
+        RTT::endlog();
+      rosnode->declare_parameter(param_name);
+    }
```

Comment:
[rclcpp::NodeOptions](http://docs.ros2.org/dashing/api/rclcpp/classrclcpp_1_1NodeOptions.html#a5b9580203b3cb88c75b5b858e26b3b7d) has a flag `allow_undeclared_parameters` that decides whether it is okay to set undeclared parameters or not. There are also good reasons to not accept this (e.g. to trigger an error in case of a typo).

I agree that it makes sense for `rtt_ros2_params` that `setParameter()` and `storeProperty()` declare new ROS parameters implicitly by default. But this should be achieved by updating the `NodeOptions` defaults in `rtt_ros2_node`, not by calling `declare_parameter()` here or in `storeProperty()` explicitly. If `rclcpp::Node::set_parameter()` fails for whatever reason, the `setParameter()` operation should also return `false`, and likewise for `storeProperty()`.

But you could add a new operation `declareParameter()` that internally calls `rclcpp::Node::declare_parameter()` in order to mimic the API as closely as possible, and to not introduce subtle differences in behavior compared to what users might now from the rclcpp or other client library APIs.